hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/CellBlocks.java

```diff
@@ -0,0 +1,427 @@
+   * Positive numbers mean the index.
+   * Otherwise (-1 * insertion point) - 1
+   */
+  private int find(Cell needle) {
+    int begin = minCellIdx;
+    int end = maxCellIdx - 1;
+
+    while (begin <= end) {
+      int mid = begin + ((end - begin) / 2);
+      Cell midCell = getCellFromIndex(mid);
+      int compareRes = comparator.compare(midCell, needle);
+
+      // 0 means equals. We found the key.
+      if (compareRes == 0) return mid;
+      else if (compareRes < 0) {
+        // midKey is less than needle so we need to look at farther up
+        begin = mid + 1;
+      } else {
+        // midKey is greater than needle so we need to look down
+        end = mid - 1;
+      }
+    }
+
+    return (-1 * begin) - 1;
+  }
+
+  @Override public Comparator<? super Cell> comparator() { return comparator; }
+
+  @Override public int size() { return maxCellIdx-minCellIdx; }
+
+  @Override public boolean isEmpty() { return (maxCellIdx==minCellIdx); }
+
+
+  // ---------------- Sub-Maps ----------------
+  @Override public ConcurrentNavigableMap<Cell, Cell> subMap( Cell fromKey,
+                                                        boolean fromInclusive,
+                                                        Cell toKey,
+                                                        boolean toInclusive) {
+    int toIndex = find(toKey);
+    if (toInclusive && toIndex >= 0) toIndex++;
+    else if (toIndex < 0) toIndex = -(toIndex + 1) - 1;
+
+    int fromIndex = find(fromKey);
+    if (!fromInclusive && fromIndex >= 0) fromIndex++;
+    else if (fromIndex < 0) fromIndex = -(fromIndex + 1);
+
+    return createCellBlocks(comparator, fromIndex, toIndex, false);
+  }
+
+  @Override public ConcurrentNavigableMap<Cell, Cell> headMap(Cell toKey, boolean inclusive) {
+    int index = find(toKey);
+    if (inclusive && index >= 0) index++;
+    else if (index < 0) index = -(index + 1) - 1;
+    return createCellBlocks(comparator, minCellIdx, index, false);
+  }
+
+  @Override public ConcurrentNavigableMap<Cell, Cell> tailMap(Cell fromKey, boolean inclusive) {
+    int index = find(fromKey);
+    if (!inclusive && index >= 0) index++;
+    else if (index < 0) index = -(index + 1);
+    return createCellBlocks(comparator, index, maxCellIdx, false);
+  }
+
+  @Override public ConcurrentNavigableMap<Cell, Cell> descendingMap() {
+    return createCellBlocks(comparator, minCellIdx, maxCellIdx, true);
+  }
+
+  @Override public ConcurrentNavigableMap<Cell, Cell> subMap(Cell k, Cell k1) {
+    return this.subMap(k, true, k1, true);
+  }
+
+  @Override 
+  public ConcurrentNavigableMap<Cell, Cell> headMap(Cell k) { return this.headMap(k, true); }
+
+  @Override 
+  public ConcurrentNavigableMap<Cell, Cell> tailMap(Cell k) { return this.tailMap(k, true); }
+
+
+  // -------------------------------- Key's getters --------------------------------
+  @Override public Cell firstKey() {
+    if (isDescending) return lastKey();
```

Comment:
Here and after. Please check the Apache coding conventions. Blocks might be required within "if -- else" cases.